lib/facilities/local_store.rb

```diff
@@ -0,0 +1,109 @@
+# frozen_string_literal: true
+require 'facilities/bulk_client'
+
+module Facilities
+  class LocalStore
+    attr_reader :adapter
+
+    GIS_CHECK_FREQUENCY = 1.hour
+    BATCH_SIZE = 1000.0
+    OPEN_TIMEOUT = 2
+    REQUEST_TIMEOUT = 6
+    Pointer = Struct.new(:x, :y, :facility)
+
+    def initialize(url, _type, adapter)
+      @url = url
+      @adapter = adapter
+      @mutex = Mutex.new
+      @last_check = Time.new(0).utc
+      @current_gis_update = 0
+      @client = BulkClient.new(@url)
+      @index = {}
+      @coordinates = []
+      # check_for_freshness
+    end
+
+    def get(id)
+      check_for_freshness
+      @mutex.synchronize do
+        return @index[id]
+      end
+    end
+
+    def query(bbox, services = nil)
+      check_for_freshness
+      filter = services ? @adapter.with_services(services) : ->(_) { true }
+      @mutex.synchronize do
+        @coordinates.select(&(within bbox)).sort_by(&(dist_from_center bbox)).map(&:facility).select(&filter)
+      end
+    end
+
+    def within(bbox)
+      x_min, x_max = min_max(bbox[0], bbox[2])
+      y_min, y_max = min_max(bbox[1], bbox[3])
+      lambda do |point|
+        return false if point.x.nil? || point.y.nil?
+        x_min < point.x && point.x < x_max &&
+          y_min < point.y && point.y < y_max
+      end
+    end
+
+    def min_max(x, y)
+      x < y ? [x, y] : [y, x]
+    end
+
+    def dist_from_center(bbox)
+      center_x = (bbox[0] + bbox[2]) / 2.0
+      center_y = (bbox[1] + bbox[3]) / 2.0
+      lambda do |point|
+        Math.sqrt((point.x - center_x)**2 + (point.y - center_y)**2)
+      end
```

Comment:
Again here the dist_from_center is really a question you want to ask the coordinates themselves, so I'd advise moving this into their class.